File: core/metrics.py
from __future__ import annotations

import bisect

from core.model import ButtonEvent, LaserNode, MeasureHoldEvent, MeasureLaserSegment, Signature, TimePoint, VoxChart


class ChartMetrics:
    def __init__(self, chart: VoxChart):
        self.chart = chart
        self._resolution = chart.beat_resolution
        self._signature_cache: dict[int, Signature] = {}
        self._measure_start_cache: dict[int, float] = {1: 0.0}
        self._measure_duration_cache: dict[int, float] = {}
        self._measure_time_segments_cache: dict[int, list[tuple[float, float, float, float]]] = {}

        # Build global BPM timeline: sorted list of (chart_units, bpm, units_per_beat)
        self._bpm_units: list[float] = []
        self._bpm_values: list[float] = []
        self._bpm_beat_units: list[float] = []
        if chart.bpms:
            for event in chart.bpms:
                self._bpm_units.append(self.chart_units(event.time))
                self._bpm_values.append(event.bpm)
                self._bpm_beat_units.append(self._resolution * 4 / event.division)
        else:
            self._bpm_units.append(0.0)
            self._bpm_values.append(120.0)
            self._bpm_beat_units.append(float(self._resolution))

    def signature_for_measure(self, measure: int) -> Signature:
        if measure in self._signature_cache:
            return self._signature_cache[measure]

        candidates = [start for start in self.chart.signatures if start <= measure]
        signature = self.chart.signatures[max(candidates)] if candidates else Signature(4, 4)
        self._signature_cache[measure] = signature
        return signature

    def local_units(self, time: TimePoint) -> float:
        signature = self.signature_for_measure(time.measure)
        return ((time.beat - 1) * signature.beat_units(self._resolution)) + time.offset

    def measure_fraction(self, time: TimePoint) -> float:
        return self.units_to_time_fraction(time.measure, self.local_units(time))

    def measure_start_units(self, measure: int) -> float:
        if measure in self._measure_start_cache:
            return self._measure_start_cache[measure]

        total = 0.0
        for current_measure in range(1, measure):
            total += self.signature_for_measure(current_measure).measure_units(self._resolution)

        self._measure_start_cache[measure] = total
        return total

    def chart_units(self, time: TimePoint) -> float:
        return self.measure_start_units(time.measure) + self.local_units(time)

    def bpm_at_units(self, abs_units: float) -> tuple[float, float]:
        idx = max(0, bisect.bisect_right(self._bpm_units, abs_units) - 1)
        return self._bpm_values[idx], self._bpm_beat_units[idx]

    def _measure_time_segments(self, measure: int) -> list[tuple[float, float, float, float]]:
        if measure in self._measure_time_segments_cache:
            return self._measure_time_segments_cache[measure]

        m_start = self.measure_start_units(measure)
        m_end = m_start + self.signature_for_measure(measure).measure_units(self._resolution)

        # Find all BPM change points within [m_start, m_end)
        lo = bisect.bisect_right(self._bpm_units, m_start) - 1
        hi = bisect.bisect_right(self._bpm_units, m_end)
        breakpoints: list[float] = [m_start]
        for i in range(max(0, lo), hi):
            u = self._bpm_units[i]
            if m_start < u < m_end:
                breakpoints.append(u)
        breakpoints.append(m_end)

        segments: list[tuple[float, float, float, float]] = []
        for a, b in zip(breakpoints, breakpoints[1:]):
            bpm, beat_units = self.bpm_at_units(a)
            local_a = a - m_start
            local_b = b - m_start
            segments.append((local_a, local_b, bpm, beat_units))

        self._measure_time_segments_cache[measure] = segments
        return segments

    def measure_duration(self, measure: int) -> float:
        if measure in self._measure_duration_cache:
            return self._measure_duration_cache[measure]

        total = 0.0
        for local_a, local_b, bpm, beat_units in self._measure_time_segments(measure):
            total += (local_b - local_a) * 60.0 / (beat_units * bpm)
        self._measure_duration_cache[measure] = total
        return total

    # Convert local measure units to a time fraction [0.0, 1.0] based on BPM changes within the measure
    def units_to_time_fraction(self, measure: int, local_units: float) -> float:
        total_duration = self.measure_duration(measure)
        if total_duration <= 0:
            return 0.0

        elapsed = 0.0
        for local_a, local_b, bpm, beat_units in self._measure_time_segments(measure):
            if local_units <= local_a:
                break
            seg_units = min(local_units, local_b) - local_a
            elapsed += seg_units * 60.0 / (beat_units * bpm)
            if local_units <= local_b:
                break
        return min(1.0, elapsed / total_duration)

    def beat_time_fractions(self, measure: int) -> list[float]:
        signature = self.signature_for_measure(measure)
        return [
            self.units_to_time_fraction(measure, beat * signature.beat_units(self._resolution))
            for beat in range(signature.beat)
        ]

    def split_hold_events(self, event: ButtonEvent) -> list[MeasureHoldEvent]:
        segments: list[MeasureHoldEvent] = []
        remaining = float(max(event.hold_length, 0))
        current_measure = event.time.measure
        current_units = self.local_units(event.time)

        while True:
            signature = self.signature_for_measure(current_measure)
            measure_units = signature.measure_units(self._resolution)
            start_fraction = self.units_to_time_fraction(current_measure, current_units)

            if remaining <= 0:
                chip_end_units = min(measure_units, current_units + 4.0)
                end_fraction = self.units_to_time_fraction(current_measure, chip_end_units)
                end_fraction = min(1.0, end_fraction)
                segments.append(
                    MeasureHoldEvent(
                        track=event.track,
                        measure=current_measure,
                        start_fraction=start_fraction,
                        end_fraction=end_fraction,
                        effect_code=event.effect_code,
                    )
                )
                return segments

            available = measure_units - current_units
            if remaining <= available:
                end_fraction = self.units_to_time_fraction(current_measure, current_units + remaining)
                segments.append(
                    MeasureHoldEvent(
                        track=event.track,
                        measure=current_measure,
                        start_fraction=start_fraction,
                        end_fraction=end_fraction,
                        effect_code=event.effect_code,
                    )
                )
                return segments

            segments.append(
                MeasureHoldEvent(
                    track=event.track,
                    measure=current_measure,
                    start_fraction=start_fraction,
                    end_fraction=1.0,
                    effect_code=event.effect_code,
                )
            )
            remaining -= available
            current_measure += 1
            current_units = 0.0

    def split_hold_segments(self, event: ButtonEvent) -> list[tuple[int, float, float]]:
        return [
            (segment.measure, segment.start_fraction, segment.end_fraction)
            for segment in self.split_hold_events(event)
        ]

    def split_laser_segments(self, nodes: list[LaserNode]) -> list[MeasureLaserSegment]:
        if len(nodes) < 2:
            return []

        segments: list[MeasureLaserSegment] = []
        for start_node, end_node in zip(nodes, nodes[1:]):
            if start_node.flag == 2 and end_node.flag == 1:
                continue

            start_units = self.chart_units(start_node.time)
            end_units = self.chart_units(end_node.time)
            if end_units < start_units:
                continue
            pair_is_horizontal = start_units == end_units

            start_measure = start_node.time.measure
            end_measure = end_node.time.measure

            for measure in range(start_measure, end_measure + 1):
                measure_start_units = self.measure_start_units(measure)
                measure_end_units = measure_start_units + self.signature_for_measure(measure).measure_units(self._resolution)
                segment_start_units = max(start_units, measure_start_units)
                segment_end_units = min(end_units, measure_end_units)

                if segment_end_units < segment_start_units:
                    continue

                if segment_end_units == segment_start_units and not pair_is_horizontal:
                    continue

                if segment_end_units > segment_start_units:
                    duration = end_units - start_units
                    start_ratio = (segment_start_units - start_units) / duration
                    end_ratio = (segment_end_units - start_units) / duration
                    start_fraction = self.units_to_time_fraction(measure, segment_start_units - measure_start_units)
                    end_fraction = self.units_to_time_fraction(measure, segment_end_units - measure_start_units)
                    start_position = start_node.position + (end_node.position - start_node.position) * start_ratio
                    end_position = start_node.position + (end_node.position - start_node.position) * end_ratio
                else:
                    start_fraction = self.units_to_time_fraction(measure, segment_start_units - measure_start_units)
                    end_fraction = start_fraction
                    start_position = start_node.position
                    end_position = end_node.position

                segments.append(
                    MeasureLaserSegment(
                        track=start_node.track,
                        measure=measure,
                        start_fraction=start_fraction,
                        end_fraction=end_fraction,
                        start_position=start_position,
                        end_position=end_position,
                        start_flag=start_node.flag,
                        end_flag=end_node.flag,
                        range_index=start_node.range_index,
                    )
                )

        return segments

File: src/metrics.c
#include "metrics.h"

#include <stdlib.h>
#include <string.h>
#include <math.h>

/* ---- helpers ---- */

double sig_beat_units(Signature sig, int resolution) {
    return (double)resolution * 4.0 / (double)sig.note;
}

double sig_measure_units(Signature sig, int resolution) {
    return (double)sig.beat * sig_beat_units(sig, resolution);
}

/* Binary search: find index of last entry in sorted array where values[i] <= key.
 * Returns -1 if key < values[0]. */
static int bisect_right_minus1(const double *values, int n, double key) {
    int lo = 0, hi = n;
    while (lo < hi) {
        int mid = (lo + hi) / 2;
        if (values[mid] <= key) lo = mid + 1;
        else hi = mid;
    }
    return lo - 1;
}

/* ---- init / free ---- */

void metrics_init(ChartMetrics *m, const VoxChart *chart) {
    memset(m, 0, sizeof(*m));
    m->chart = chart;
    m->resolution = chart->beat_resolution;

    /* Build BPM timeline */
    if (chart->num_bpms > 0) {
        m->num_bpm_entries = chart->num_bpms;
    } else {
        m->num_bpm_entries = 1;
    }
    m->bpm_units      = (double *)calloc((size_t)m->num_bpm_entries, sizeof(double));
    m->bpm_values     = (double *)calloc((size_t)m->num_bpm_entries, sizeof(double));
    m->bpm_beat_units = (double *)calloc((size_t)m->num_bpm_entries, sizeof(double));

    if (chart->num_bpms > 0) {
        /* We must compute chart_units for BPM events which requires the signature info.
         * Build a temporary local ChartMetrics with just signature data (no BPM cache)
         * to compute measure_start_units without the BPM cache. */
        /* For now, store only the measure/beat/offset -> chart_units mapping.
         * We use a forward pass since signatures don't depend on BPM. */

        /* accumulate measure start units from scratch for BPM events */
        /* Build measure_start_units inline for the BPM timeline build */
        int max_bpm_measure = 0;
        for (int i = 0; i < chart->num_bpms; i++) {
            if (chart->bpms[i].time.measure > max_bpm_measure)
                max_bpm_measure = chart->bpms[i].time.measure;
        }

        /* Compute chart_units for each BPM event:
         * We walk measure by measure accumulating units. */
        /* Simple O(M * B) pass: for each BPM event compute its chart_units. */
        /* This is called once so O(M) is fine. */
        for (int i = 0; i < chart->num_bpms; i++) {
            TimePoint tp = chart->bpms[i].time;
            /* compute measure_start_units(tp.measure) fresh */
            double mstart = 0.0;
            for (int mu = 1; mu < tp.measure; mu++) {
                /* find sig for mu */
                Signature s = {4, 4};
                int best = -1;
                for (int si = 0; si < chart->num_sigs; si++) {
                    if (chart->sig_measure[si] <= mu) {
                        if (best < 0 || chart->sig_measure[si] > chart->sig_measure[best])
                            best = si;
                    }
                }
                if (best >= 0) s = chart->sig_value[best];
                mstart += sig_measure_units(s, m->resolution);
            }
            /* find sig for tp.measure */
            Signature sig = {4, 4};
            {
                int best = -1;
                for (int si = 0; si < chart->num_sigs; si++) {
                    if (chart->sig_measure[si] <= tp.measure) {
                        if (best < 0 || chart->sig_measure[si] > chart->sig_measure[best])
                            best = si;
                    }
                }
                if (best >= 0) sig = chart->sig_value[best];
            }
            double local = ((double)(tp.beat - 1)) * sig_beat_units(sig, m->resolution) + (double)tp.offset;
            m->bpm_units[i] = mstart + local;
            m->bpm_values[i] = chart->bpms[i].bpm;
            m->bpm_beat_units[i] = (double)m->resolution * 4.0 / (double)chart->bpms[i].division;
        }
    } else {
        m->bpm_units[0]      = 0.0;
        m->bpm_values[0]     = 120.0;
        m->bpm_beat_units[0] = (double)m->resolution;
    }

    /* Allocate measure start cache up front for a reasonable max */
    m->cache_size = 4096;
    m->measure_start_cache   = (double *)malloc((size_t)m->cache_size * sizeof(double));
    m->measure_duration_cache = (double *)malloc((size_t)m->cache_size * sizeof(double));
    for (int i = 0; i < m->cache_size; i++) {
        m->measure_start_cache[i]    = -1.0;
        m->measure_duration_cache[i] = -1.0;
    }
    m->measure_start_cache[1] = 0.0; /* measure 1 starts at 0 */
}

void metrics_free(ChartMetrics *m) {
    free(m->bpm_units);
    free(m->bpm_values);
    free(m->bpm_beat_units);
    free(m->measure_start_cache);
    free(m->measure_duration_cache);
}

/* ---- signature lookup ---- */

Signature metrics_signature(const ChartMetrics *m, int measure) {
    Signature result = {4, 4};
    int best_measure = -1;
    for (int i = 0; i < m->chart->num_sigs; i++) {
        int sm = m->chart->sig_measure[i];
        if (sm <= measure && sm > best_measure) {
            best_measure = sm;
            result = m->chart->sig_value[i];
        }
    }
    return result;
}

/* ---- measure start units (cached) ---- */

double metrics_measure_start_units(ChartMetrics *m, int measure) {
    if (measure < 1) return 0.0;

    /* grow cache if needed */
    if (measure >= m->cache_size) {
        int new_size = m->cache_size;
        while (new_size <= measure) new_size *= 2;
        m->measure_start_cache    = (double *)realloc(m->measure_start_cache, (size_t)new_size * sizeof(double));
        m->measure_duration_cache = (double *)realloc(m->measure_duration_cache, (size_t)new_size * sizeof(double));
        for (int i = m->cache_size; i < new_size; i++) {
            m->measure_start_cache[i]    = -1.0;
            m->measure_duration_cache[i] = -1.0;
        }
        m->cache_size = new_size;
    }

    if (m->measure_start_cache[measure] >= 0.0) {
        return m->measure_start_cache[measure];
    }

    /* compute iteratively from last cached measure */
    int start = 1;
    while (start < measure && m->measure_start_cache[start] >= 0.0) start++;
    /* start is the first measure we need to compute */
    double total = (start > 1) ? m->measure_start_cache[start - 1] : 0.0;
    /* add sig_measure_units for each measure from (start-1) up */
    for (int cur = start - 1; cur < measure; cur++) {
        if (cur < 1) { total = 0.0; continue; }
        Signature sig = metrics_signature(m, cur);
        double mu = sig_measure_units(sig, m->resolution);
        total += mu;
        if (cur + 1 < m->cache_size)
            m->measure_start_cache[cur + 1] = total;
    }

    m->measure_start_cache[measure] = total;
    return total;
}

/* ---- local units ---- */

double metrics_local_units(ChartMetrics *m, TimePoint tp) {
    Signature sig = metrics_signature(m, tp.measure);
    return (double)(tp.beat - 1) * sig_beat_units(sig, m->resolution) + (double)tp.offset;
}

double metrics_chart_units(ChartMetrics *m, TimePoint tp) {
    return metrics_measure_start_units(m, tp.measure) + metrics_local_units(m, tp);
}

/* ---- BPM at a given absolute chart_units ---- */

static void bpm_at_units(const ChartMetrics *m, double abs_units,
                         double *out_bpm, double *out_beat_units) {
    int idx = bisect_right_minus1(m->bpm_units, m->num_bpm_entries, abs_units);
    if (idx < 0) idx = 0;
    *out_bpm        = m->bpm_values[idx];
    *out_beat_units = m->bpm_beat_units[idx];
}

/* ---- measure time segments ---- */

/* A segment: [local_a, local_b) within the measure, with constant BPM. */
typedef struct { double local_a, local_b, bpm, beat_units; } MeasureSegment;

/* Fill segments[] for measure; returns count. */
static int get_measure_segments(ChartMetrics *m, int measure,
                                MeasureSegment *segs, int cap) {
    double m_start = metrics_measure_start_units(m, measure);
    Signature sig = metrics_signature(m, measure);
    double m_dur = sig_measure_units(sig, m->resolution);
    double m_end = m_start + m_dur;

    /* collect BPM change points inside (m_start, m_end) */
    double bp[256];
    int nbp = 0;
    bp[nbp++] = m_start;

    int lo = bisect_right_minus1(m->bpm_units, m->num_bpm_entries, m_start);
    if (lo < 0) lo = 0;
    /* include lo since it might equal m_start */
    for (int i = lo; i < m->num_bpm_entries && m->bpm_units[i] <= m_end; i++) {
        double u = m->bpm_units[i];
        if (u > m_start && u < m_end) {
            if (nbp < 255) bp[nbp++] = u;
        }
    }
    bp[nbp++] = m_end;

    int n = 0;
    for (int i = 0; i < nbp - 1 && n < cap; i++) {
        double a = bp[i], b = bp[i+1];
        double bpm, beat_u;
        bpm_at_units(m, a, &bpm, &beat_u);
        segs[n].local_a     = a - m_start;
        segs[n].local_b     = b - m_start;
        segs[n].bpm         = bpm;
        segs[n].beat_units  = beat_u;
        n++;
    }
    return n;
}

/* ---- measure duration ---- */

double metrics_measure_duration(ChartMetrics *m, int measure) {
    if (measure >= 1 && measure < m->cache_size && m->measure_duration_cache[measure] >= 0.0) {
        return m->measure_duration_cache[measure];
    }

    MeasureSegment segs[256];
    int n = get_measure_segments(m, measure, segs, 256);
    double total = 0.0;
    for (int i = 0; i < n; i++) {
        total += (segs[i].local_b - segs[i].local_a) * 60.0 / (segs[i].beat_units * segs[i].bpm);
    }
    if (measure >= 1 && measure < m->cache_size)
        m->measure_duration_cache[measure] = total;
    return total;
}

/* ---- units to time fraction ---- */

double metrics_units_to_frac(ChartMetrics *m, int measure, double local_units) {
    double total_dur = metrics_measure_duration(m, measure);
    if (total_dur <= 0.0) return 0.0;

    MeasureSegment segs[256];
    int n = get_measure_segments(m, measure, segs, 256);

    double elapsed = 0.0;
    for (int i = 0; i < n; i++) {
        double la = segs[i].local_a, lb = segs[i].local_b;
        if (local_units <= la) break;
        double seg_u = (local_units < lb ? local_units : lb) - la;
        elapsed += seg_u * 60.0 / (segs[i].beat_units * segs[i].bpm);
        if (local_units <= lb) break;
    }
    double frac = elapsed / total_dur;
    return frac < 1.0 ? frac : 1.0;
}

double metrics_measure_frac(ChartMetrics *m, TimePoint tp) {
    return metrics_units_to_frac(m, tp.measure, metrics_local_units(m, tp));
}

/* ---- beat fractions ---- */

int metrics_beat_fracs(ChartMetrics *m, int measure, double *out, int out_cap) {
    Signature sig = metrics_signature(m, measure);
    int n = sig.beat < out_cap ? sig.beat : out_cap;
    double bu = sig_beat_units(sig, m->resolution);
    for (int i = 0; i < n; i++) {
        out[i] = metrics_units_to_frac(m, measure, (double)i * bu);
    }
    return n;
}

/* ---- split hold events ---- */

int metrics_split_hold(ChartMetrics *m, const ButtonEvent *ev,
                       MeasureHoldEvent *out_segs, int out_cap) {
    int n = 0;
    double remaining = (double)(ev->hold_length > 0 ? ev->hold_length : 0);
    int cur_measure = ev->time.measure;
    double cur_units = metrics_local_units(m, ev->time);

    while (n < out_cap) {
        Signature sig = metrics_signature(m, cur_measure);
        double measure_units_total = sig_measure_units(sig, m->resolution);
        double start_frac = metrics_units_to_frac(m, cur_measure, cur_units);

        if (remaining <= 0.0) {
            double chip_end = cur_units + 4.0;
            if (chip_end > measure_units_total) chip_end = measure_units_total;
            double end_frac = metrics_units_to_frac(m, cur_measure, chip_end);
            if (end_frac > 1.0) end_frac = 1.0;
            out_segs[n].track       = ev->track;
            out_segs[n].measure     = cur_measure;
            out_segs[n].start_frac  = start_frac;
            out_segs[n].end_frac    = end_frac;
            out_segs[n].effect_code = ev->effect_code;
            n++;
            return n;
        }

        double available = measure_units_total - cur_units;
        if (remaining <= available) {
            double end_frac = metrics_units_to_frac(m, cur_measure, cur_units + remaining);
            out_segs[n].track       = ev->track;
            out_segs[n].measure     = cur_measure;
            out_segs[n].start_frac  = start_frac;
            out_segs[n].end_frac    = end_frac;
            out_segs[n].effect_code = ev->effect_code;
            n++;
            return n;
        }

        out_segs[n].track       = ev->track;
        out_segs[n].measure     = cur_measure;
        out_segs[n].start_frac  = start_frac;
        out_segs[n].end_frac    = 1.0;
        out_segs[n].effect_code = ev->effect_code;
        n++;
        remaining -= available;
        cur_measure++;
        cur_units = 0.0;
    }
    return n;
}

/* ---- split laser segments ---- */

int metrics_split_laser(ChartMetrics *m,
                        const LaserNode *nodes, int n_nodes,
                        MeasureLaserSegment *out_segs, int out_cap) {
    int n_out = 0;

    for (int i = 0; i + 1 < n_nodes && n_out < out_cap; i++) {
        const LaserNode *sn = &nodes[i];
        const LaserNode *en = &nodes[i + 1];

        if (sn->flag == 2 && en->flag == 1) continue;

        double start_units = metrics_chart_units(m, sn->time);
        double end_units   = metrics_chart_units(m, en->time);
        if (end_units < start_units) continue;
        int pair_is_horizontal = (start_units == end_units);

        int start_measure = sn->time.measure;
        int end_measure   = en->time.measure;

        for (int measure = start_measure; measure <= end_measure && n_out < out_cap; measure++) {
            double ms_units = metrics_measure_start_units(m, measure);
            Signature sig = metrics_signature(m, measure);
            double me_units = ms_units + sig_measure_units(sig, m->resolution);

            double seg_start_u = start_units > ms_units ? start_units : ms_units;
            double seg_end_u   = end_units   < me_units ? end_units   : me_units;

            if (seg_end_u < seg_start_u) continue;
            if (seg_end_u == seg_start_u && !pair_is_horizontal) continue;

            MeasureLaserSegment *seg = &out_segs[n_out];
            seg->track      = sn->track;
            seg->measure    = measure;
            seg->start_flag = sn->flag;
            seg->end_flag   = en->flag;
            seg->range_index = sn->range_index;

            if (seg_end_u > seg_start_u) {
                double duration   = end_units - start_units;
                double start_ratio = (seg_start_u - start_units) / duration;
                double end_ratio   = (seg_end_u   - start_units) / duration;
                seg->start_frac = metrics_units_to_frac(m, measure, seg_start_u - ms_units);
                seg->end_frac   = metrics_units_to_frac(m, measure, seg_end_u   - ms_units);
                seg->start_pos  = sn->position + (en->position - sn->position) * start_ratio;
                seg->end_pos    = sn->position + (en->position - sn->position) * end_ratio;
            } else {
                /* horizontal segment */
                seg->start_frac = metrics_units_to_frac(m, measure, seg_start_u - ms_units);
                seg->end_frac   = seg->start_frac;
                seg->start_pos  = sn->position;
                seg->end_pos    = en->position;
            }
            n_out++;
        }
    }
    return n_out;
}
